lcli/src/rescue.rs

```diff
@@ -0,0 +1,128 @@
+use clap::ArgMatches;
+use clap_utils::parse_required;
+use environment::Environment;
+use eth2::{
+    types::{BlockId, ChainSpec, ForkName, PublishBlockRequest, SignedBlockContents},
+    BeaconNodeHttpClient, SensitiveUrl, Timeouts,
+};
+use eth2_network_config::Eth2NetworkConfig;
+use ssz::Encode;
+use std::fs::File;
+use std::io::{Read, Write};
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::Duration;
+use types::EthSpec;
+
+const HTTP_TIMEOUT: Duration = Duration::from_secs(3600);
+
+pub fn run<T: EthSpec>(
+    env: Environment<T>,
+    network_config: Eth2NetworkConfig,
+    matches: &ArgMatches,
+) -> Result<(), String> {
+    let executor = env.core_context().executor;
+    executor
+        .handle()
+        .ok_or("shutdown in progress")?
+        .block_on(async move { run_async::<T>(network_config, matches).await })
+        .unwrap();
+    Ok(())
+}
+
+pub async fn run_async<T: EthSpec>(
+    network_config: Eth2NetworkConfig,
+    matches: &ArgMatches,
+) -> Result<(), String> {
+    let spec = &network_config.chain_spec::<T>()?;
+    let source_url: SensitiveUrl = parse_required(matches, "source-url")?;
+    let target_url: SensitiveUrl = parse_required(matches, "target-url")?;
+    let start_block: BlockId = parse_required(matches, "start-block")?;
+
+    let source = BeaconNodeHttpClient::new(source_url, Timeouts::set_all(HTTP_TIMEOUT));
+    let target = BeaconNodeHttpClient::new(target_url, Timeouts::set_all(HTTP_TIMEOUT));
+
+    // 1. Download blocks back from head, looking for common ancestor.
+    let mut blocks = vec![];
+    let mut next_block_id = start_block;
+    loop {
+        println!("downloading {next_block_id:?}");
+
+        let publish_block_req = get_block_from_source::<T>(&source, next_block_id, spec).await;
+        let block = publish_block_req.signed_block();
+
+        next_block_id = BlockId::Root(block.parent_root());
+        blocks.push((block.slot(), publish_block_req));
+
+        let block_exists_in_target = target
+            .get_beacon_blocks_ssz::<T>(next_block_id, spec)
+            .await
+            .unwrap()
+            .is_some();
+        if block_exists_in_target {
+            println!("common ancestor found: {next_block_id:?}");
+            break;
+        }
+    }
+
+    // 2. Apply blocks to target.
+    for (slot, block) in blocks.iter().rev() {
+        println!("posting block at slot {slot}");
+        if let Err(e) = target.post_beacon_blocks(block).await {
+            println!("error posting {slot}: {e:?}");
```

Comment:
If this is a non 200 status code, we should just break out of the loop.

Reply:
it could be a 202 (or other error) for block-is-already-known

Reply:
but actually, that probably shouldn't happen, so yeah, I agree

Reply:
Yep It was a 202, I added a check for this